File: python/cicada_nn/multi_timeframe.py
"""Map LTF bar timestamps to HTF bar indices for real multi-timeframe signals."""

from __future__ import annotations

HIGHER_TIMEFRAME: dict[str, str | None] = {
    "M1": "M5",
    "M5": "M15",
    "M15": "H1",
    "M30": "H1",
    "H1": "H4",
    "H4": "D1",
    "D1": "W1",
    "W1": None,
}


def get_higher_timeframe(tf: str) -> str | None:
    return HIGHER_TIMEFRAME.get(tf.upper(), None)


def build_htf_index_for_each_ltf_bar(
    ltf_bars: list[dict], htf_bars: list[dict]
) -> list[int]:
    """For each LTF bar index i, HTF bar index j with htf_bars[j]['time'] <= ltf_bars[i]['time'] (latest such j)."""
    if not htf_bars:
        return [-1] * len(ltf_bars)
    out: list[int] = []
    for lb in ltf_bars:
        t = float(lb.get("time", 0))
        lo, hi = 0, len(htf_bars) - 1
        best = -1
        while lo <= hi:
            mid = (lo + hi) // 2
            mt = float(htf_bars[mid].get("time", 0))
            if mt <= t:
                best = mid
                lo = mid + 1
            else:
                hi = mid - 1
        out.append(best)
    return out

File: src/app/core/multiTimeframe.ts
/**
 * Higher timeframe (HTF) mapping and LTF↔HTF bar alignment for real multi-timeframe signals.
 * HTF bars must be the same symbol, sorted by time ascending (open time).
 */

import type { OHLCVBar } from './ohlcv';
import { getBarDurationMs } from './ohlcv';
import type { Timeframe } from './types';

/** Next higher timeframe for bias / filter (institutional: trade LTF, filter with HTF). */
export const HIGHER_TIMEFRAME: Record<Timeframe, Timeframe | null> = {
  M1: 'M5',
  M5: 'M15',
  M15: 'H1',
  M30: 'H1',
  H1: 'H4',
  H4: 'D1',
  D1: 'W1',
  W1: null,
};

export function getHigherTimeframe(tf: Timeframe): Timeframe | null {
  return HIGHER_TIMEFRAME[tf] ?? null;
}

/**
 * For each LTF bar, index of the HTF bar whose open time is ≤ LTF bar time (current HTF candle).
 * Returns -1 if no HTF bar is at or before this time.
 */
export function buildHtfIndexForEachLtfBar(ltfBars: OHLCVBar[], htfBars: OHLCVBar[]): Int32Array {
  const out = new Int32Array(ltfBars.length);
  if (!htfBars.length) {
    out.fill(-1);
    return out;
  }
  for (let i = 0; i < ltfBars.length; i++) {
    const t = ltfBars[i].time;
    let lo = 0;
    let hi = htfBars.length - 1;
    let best = -1;
    while (lo <= hi) {
      const mid = (lo + hi) >> 1;
      if (htfBars[mid].time <= t) {
        best = mid;
        lo = mid + 1;
      } else {
        hi = mid - 1;
      }
    }
    out[i] = best;
  }
  return out;
}

export function barDurationMs(tf: Timeframe): number {
  return getBarDurationMs(tf);
}
